Modules/_threadmodule.c

```diff
@@ -1921,13 +2037,99 @@ PyDoc_STRVAR(thread__is_main_interpreter_doc,
 \n\
 Return True if the current interpreter is the main Python interpreter.");
 
+static PyObject *
+thread_shutdown(PyObject *self, PyObject *args)
+{
+    PyThread_ident_t ident = PyThread_get_thread_ident_ex();
+    thread_module_state *state = get_thread_state(self);
+
+    for (;;) {
```

Comment:
This looks like the loop used in the nogil forks, but I think we can do something simpler because the handles are in their own linked list:

Move the linked list to a local variable (clearing `state->shutdown_handles`) and then process the local linked list:

Roughly:

```c
HEAD_LOCK(&_PyRuntime);
struct llist_node list = LLIST_INIT(list);
list_concat(&list, &state->shutdown_handles);
HEAD_UNLOCK(&PyRuntime);

struct llist_node *node;
llist_for_each(node, &list) {
    ...
}
```

Reply:
I don't think this ends up simplifying things. We'd still need to iterate over `list` while the lock is held, because the threads that are currently running may end up removing themselves.